File: Models_stability_plasticity_dilemma/Matlab_code/RBM_synaptic.m

%% Defining amount of loops
Rep=10;                         %amount of replications
iterations=5;                   %number of iterations in negative phase
Tr=2400;                        %number of trials
betas=11;                       %beta iterations
beta=0:1/(betas-1):1;           %learning rate values

%other variables
POT_1=Tr/3;                     %point of switch to task rule 2
POT_2=2*Tr/3;                   %switch again to task rule 1
part1=1:POT_1;                  %first part of task
part2=POT_1+1:POT_2;            %second part of task
part3=POT_2+1:Tr;               %third part of task
%% model build-up
nStim=6;                        %number of units in input layer
nM1=4;                          %number of units in hidden layer
nResp=2;                        %number of units in output layer
nInput=8;                       %number of input patterns
bias=5;                         %bias parameter

%% In- and output patterns
%Input patterns
Activation=zeros(nStim,nInput);
Activation(1:6,1)=[1,0,1,0,1,0];              %blauw links cirkel    (1)
Activation(1:6,2)=[1,0,0,1,1,0];              %blauw rechts cirkel   (2)
Activation(1:6,3)=[0,1,1,0,1,0];              %rood links cirkel     (3)
Activation(1:6,4)=[0,1,0,1,1,0];              %rood rechts cirkel    (4)
Activation(1:6,5)=[1,0,1,0,0,1];              %blauw links vierkant  (5)      
Activation(1:6,6)=[1,0,0,1,0,1];              %blauw rechts vierkant (6)   
Activation(1:6,7)=[0,1,1,0,0,1];              %rood links vierkant   (7)
Activation(1:6,8)=[0,1,0,1,0,1];              %rood rechts vierkant  (8)

%learning objectives for output layer
Objective=zeros(nResp,nInput,Tr);    
Objective(1,[1,2,5,7],part1)=1;
Objective(2,[3,4,6,8],part1)=1;
Objective(1,[1,2,5,7],part3)=1;
Objective(2,[3,4,6,8],part3)=1;
Objective(1,[3,4,6,8],part2)=1;
Objective(2,[1,2,5,7],part2)=1;

%% simulation loops
for b=1:betas
    for r=1:Rep
        
%% model initialization
%Processing module
%positive phase
Rate_Input_plus=zeros(nStim,Tr);                   %Input neurons positive phase
Rate_M1_plus=zeros(nM1,Tr);                        %Hidden neurons positive phase
Rate_Out_plus=zeros(nResp,Tr);                     %Output neurons positive phase

net_M1_plus=zeros(nM1,Tr);                         %net input hidden layer positive phase

Rate_Input_min=zeros(nStim,Tr,iterations);                    %Input neurons negative phase
Rate_M1_min=zeros(nM1,Tr,iterations);            %Hidden neurons negative phase
Rate_Out_min=zeros(nResp,Tr,iterations);         %Output neurons negative phase

net_M1_min=zeros(nM1,Tr,iterations);               %net input hidden layer negative phase
net_Out_min=zeros(nResp,Tr,iterations);            %net input output layer negative phase

%binarizations
bin_M1=zeros(nM1,Tr,iterations);

%weights
W_IM1=zeros(nStim,nM1,Tr);                         %weights between input and hidden layer
W_M1O=zeros(nM1,nResp,Tr);                         %weights between hidden and output layer
%starting values for weights
W_IM1(:,:,1)=rand(nStim,nM1)*5;        
W_M1O(:,:,1)=rand(nM1,nResp)*5;  

%% Learning

%weight change
delta_M1=zeros(nStim,nM1,Tr);       
delta_M1out=zeros(nM1,nResp,Tr); 
    
Errorscore=zeros(nResp,Tr);         %errorscore at negative phase

%% Input to model
In=repmat(1:nInput,Tr/nInput);  %make list of input patterns
In=In(1,:);                                     

Input=zeros(1,Tr);
Input(1,:)=In(1,randperm(Tr));             %randomization of these input list

%% recording
Z=zeros(nStim,Tr);
Q=zeros(nResp,Tr);
Q_min=zeros(nResp,Tr,iterations);
response=zeros(nResp,Tr,iterations);
rew=zeros(1,Tr);                   

%% Model
for trial=1:Tr
    
    %Positive phase
    Z(:,trial)=Activation(:,Input(1,trial));
    Q(:,trial)=Objective(:,Input(1,trial),trial);
    %clamping visible layers
    Rate_Input_plus(:,trial)=Z(:,trial);
    Rate_Out_plus(:,trial)=Q(:,trial);
    
    %Hidden activation
    net_M1_plus(:,trial)=W_IM1(:,:,trial)'*Rate_Input_plus(:,trial)+ W_M1O(:,:,trial) *Rate_Out_plus(:,trial)-bias; %input from both layers
    Rate_M1_plus(:,trial)=ones(nM1,1)./(ones(nM1,1)+exp(-net_M1_plus(:,trial)));
    
    %Negative phase
    for i=1:iterations
        Rate_Input_min(:,trial,i)=Z(:,trial);
    
        net_M1_min(:,trial,i)=W_IM1(:,:,trial)'*Rate_Input_min(:,trial)+ W_M1O(:,:,trial) *Q_min(:,trial,i)-bias;
        Rate_M1_min(:,trial,i)=ones(nM1,1)./(ones(nM1,1)+exp(-net_M1_min(:,trial,i)));
        
        bin_M1(:,trial,i)=Rate_M1_min(:,trial,i)>rand(nM1,1,1);
        
        net_Out_min(:,trial,i)=W_M1O(:,:,trial)'*bin_M1(:,trial,i)-bias;
        Rate_Out_min(:,trial,i)=ones(nResp,1)./(ones(nResp,1)+exp(-net_Out_min(:,trial,i)));
        
        %choose biggest 
        if Rate_Out_min(1,trial,i)>Rate_Out_min(2,trial,i)
            response(1,trial,i)=1;
        else
            response(2,trial,i)=1;
        end;
        Q_min(:,trial,i+1)=response(:,trial,i);

    end;
    
    if response(:,trial,i)== Q(:,trial)  
        rew(1,trial)=1;
    else
        rew(1,trial)=0;
    end;
    
    %compute errorscore at each output unit
    Errorscore(:,trial)=(Rate_Out_plus(:,trial)-squeeze(Rate_Out_min(:,trial,i))).^2;
    
    %compute weightchange
    delta_M1(:,:,trial+1)=beta(1,b) *((Rate_Input_plus(:,trial)*Rate_M1_plus(:,trial)')-(Rate_Input_min(:,trial,i)*squeeze(Rate_M1_min(:,trial,i))'));
    delta_M1out(:,:,trial+1)=beta(1,b) * ((Rate_M1_plus(:,trial)*Rate_Out_plus(:,trial)')-(Rate_M1_min(:,trial,i)*Rate_Out_min(:,trial,i)'));
    %change weights
    W_IM1(:,:,trial+1)=W_IM1(:,:,trial) + delta_M1(:,:,trial+1);
    W_M1O(:,:,trial+1)= W_M1O(:,:,trial) + delta_M1out(:,:,trial+1);
    
    prog=trial
end;

%% track learning by accuracy over bins
 nbins=20;
 binned_Errorscore_min=zeros(1,nbins*3);
 binned_accuracy_min=zeros(1,nbins*3);
 binned_Errorscore_test=zeros(1,nbins*3);
 binned_accuracy_test=zeros(1,nbins*3);

 bin_edges=zeros(1,(nbins*3)+1);
 
 bin_edges(1,1:nbins+1)=0:POT_1/nbins:POT_1;
 bin_edges(1,nbins+1:(nbins*2)+1)=POT_1:POT_1/nbins:POT_2;
 bin_edges(1,(2*nbins)+1:(nbins*3)+1)=POT_2:POT_1/nbins:Tr; 
  
 for bin=1:nbins*3
     binned_Errorscore_min(1,bin)=mean(mean(Errorscore(:,(bin_edges(bin)+1):bin_edges(bin+1))));
     binned_accuracy_min(1,bin)=mean(rew(1,(bin_edges(bin)+1):bin_edges(bin+1)));
 end;
 save(['RBM_nosync_Beta',num2str(b),'Rep',num2str(r)],'binned_Errorscore_min','binned_accuracy_min');
    end;
end;
    

File: Models_stability_plasticity_dilemma/biorxiv2018/Python_code/RBM_synaptic.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Mar  8 14:15:08 2018

@author: pieter
"""

import numpy as np
import pickle

# Defining amount of loops
Rep=10                          # amount of replications
iterations=5                    # number of iterations in negative phase
Tr=2400                         # amount of trials
betas=11                        # beta iterations
Beta=np.arange(0,1.1,0.1)       #learning rate values

# other variables
POT_1=Tr//3                     #point of switch to task rule 2 (trial 20)
POT_2=2*Tr//3                   #switch again to task rule 1    (trial 40)
part1=np.arange(POT_1)          #first part
part2=np.arange(POT_1,POT_2)    #second part
part3=np.arange(POT_2,Tr)       #third part

#model build-up
    #Processing module
nStim=6                         #number of input units
nM1=4                           #number of hidden units in module 1
nResp=2                         #number of response options
nInput=(nStim//3)**3            #number of Input patterns
bias=5                          #bias parameter

#Input patterns
Activation=np.zeros((nStim,nInput))
Activation[:,0]=[1,0,1,0,1,0]   #blue left circle
Activation[:,1]=[1,0,0,1,1,0]   #blue right circle
Activation[:,2]=[0,1,1,0,1,0]   #red left circle
Activation[:,3]=[0,1,0,1,1,0]   #red right circle
Activation[:,4]=[1,0,1,0,0,1]   #blue left square
Activation[:,5]=[1,0,0,1,0,1]   #blue right square
Activation[:,6]=[0,1,1,0,0,1]   #red left square
Activation[:,7]=[0,1,0,1,0,1]   #red right square

#learning objectives
objective=np.zeros((nResp,nInput,Tr)) 
objective[0,[[0],[1],[4],[6]],part1]=1
objective[1,[[2],[3],[5],[7]],part1]=1
objective[1,[[0],[1],[4],[6]],part2]=1
objective[0,[[2],[3],[5],[7]],part2]=1
objective[0,[[0],[1],[4],[6]],part3]=1
objective[1,[[2],[3],[5],[7]],part3]=1

# simulation loops
for b in range(10,11):          #gradual change of beta 
    for r in range(1):          #replication loop

        # model initialization
        #processing module
        #positive phase
        Rate_Input_plus=np.zeros((nStim,Tr))                    #rate neurons
        
        Rate_M1_plus=np.zeros((nM1,Tr))                         #rate neurons        
        
        Rate_Out_plus=np.zeros((nResp,Tr))                      #rate neurons
        
        #net inputs
        net_M1_plus=np.zeros((nM1,Tr))
        
        #negative phase
        Rate_Input_min=np.zeros((nStim,Tr,iterations))       #rate neurons
        
        Rate_M1_min=np.zeros((nM1,Tr,iterations))            #rate neurons         
        
        Rate_Out_min=np.zeros((nResp,Tr,iterations))         #rate neurons
        
        #net inputs
        net_M1_min=np.zeros((nM1,Tr,iterations))
        net_Out_min=np.zeros((nResp,Tr,iterations))
        
        #binarizations
        bin_M1=np.zeros((nM1,Tr,iterations))
        
        #weights
        W_IM1=np.zeros((nStim,nM1,Tr+1))
        W_M1O=np.zeros((nM1,nResp,Tr+1))
        #initial (random) weigth strengths
        W_IM1[:,:,0]=np.random.random((nStim,nM1))*5 
        W_M1O[:,:,0]=np.random.random((nM1,nResp))*5 

        #learning
        Errorscore=np.zeros((nResp,Tr))                     #errorscore
        delta_M1out=np.zeros((nM1,nResp,Tr))                #delta hidden to output layer
        delta_M1=np.zeros((nStim,nM1,Tr));                  #delta input to hidden layer M1
        
        # Input
        #randomization of input patterns
        In=np.tile(range(nInput),(3,POT_1))
        Input=np.zeros((1,Tr))
        Input[0,part1]=In[0,np.random.permutation(POT_1)]
        Input[0,part2]=In[1,np.random.permutation(POT_1)]
        Input[0,part3]=In[2,np.random.permutation(POT_1)]
        
        #recordings
        Z=np.zeros((nStim,Tr))                              #input matrix
        Q=np.zeros((nResp,Tr))
        Q_min=np.zeros((nResp,Tr,iterations+1))
        response=np.zeros((nResp,Tr,iterations))           #response record 
        rew=np.zeros((1,Tr))                               #reward
        
        # the model

        for trial in range(Tr):            #trial loop
    
            #input and output
            Z[:,trial]=Activation[:,int(Input[0,trial])]
            Q[:,trial]=objective[:,int(Input[0,trial]),trial]
                        
            #updating rate code units
            Rate_Input_plus[:,trial]=Z[:,trial] 
                
            Rate_Out_plus[:,trial]=Q[:,trial]
                
            net_M1_plus[:,trial]=np.matmul(np.transpose(W_IM1[:,:,trial]),Rate_Input_plus[:,trial])+np.matmul(W_M1O[:,:,trial],Rate_Out_plus[:,trial])-bias
            Rate_M1_plus[:,trial]=1/(1+np.exp(-net_M1_plus[:,trial]))
            
            #negative phase
            for i in range(iterations):
                #from input and output to hidden layer
                #updating rate code units
                Rate_Input_min[:,trial,i]=Z[:,trial]
                
                net_M1_min[:,trial,i]=np.matmul(np.transpose(W_IM1[:,:,trial]),Rate_Input_min[:,trial,i])+np.matmul(W_M1O[:,:,trial],Q_min[:,trial,i])-bias
                Rate_M1_min[:,trial,i]=1/(1+np.exp(-net_M1_min[:,trial,i])) 
                
                #binarization of hidden layer activation
                bin_M1[:,trial,i]=(Rate_M1_min[:,trial,i]>np.random.random()).astype(int)
            
                #from hidden back to output
                net_Out_min[:,trial,i]=np.matmul(np.transpose(W_M1O[:,:,trial]),bin_M1[:,trial,i]) -bias
                Rate_Out_min[:,trial,i]=1/(1+np.exp(-net_Out_min[:,trial,i]))
            
                #response determination
                if Rate_Out_min[0,trial,i]>Rate_Out_min[1,trial,i]:
                    response[0,trial,i]=1
                else:
                    response[1,trial,i]=1
            
                Q_min[:,trial,i+1]=response[:,trial,i]

            #reward value determination
            if np.all(response[:,trial,i]==Q[:,trial]):
                rew[0,trial]=1
            else:
                rew[0,trial]=0 
            
            #Weight updating
            #compute general errorscore at each output unit
            Errorscore[:,trial]=(objective[:,int(Input[0,trial]),trial]-(Rate_Out_min[:,trial,i-1]))**2
            #compute delta_output
            delta_M1out[:,:,trial]=np.matmul(Rate_M1_plus[:,trial][:,None],Rate_Out_plus[:,trial][None,:])-np.matmul(Rate_M1_min[:,trial,i][:,None],Rate_Out_min[:,trial,i-1][None,:])
            #update weights from hidden M1 to output layer
            W_M1O[:,:,trial+1]=W_M1O[:,:,trial] + Beta[b]* delta_M1out[:,:,trial] 
            #compute delta hidden layer M1
            delta_M1[:,:,trial]=np.matmul(Rate_Input_plus[:,trial][:,None],Rate_M1_plus[:,trial][None,:])- np.matmul(Rate_Input_min[:,trial,i][:,None],Rate_M1_min[:,trial,i][None,:])
            #update weights from input to hidden layer M1
            W_IM1[:,:,trial+1]=W_IM1[:,:,trial]+ Beta[b] * delta_M1[:,:,trial]

            print(trial) 
         
        nbins=20
        binned_Errorscore=np.zeros((1,nbins*3))
        binned_accuracy=np.zeros((1,nbins*3))
        bin_edges=np.zeros((1,(nbins*3)+1))
 
        bin_edges[0,range(nbins)]=np.arange(0,POT_1,POT_1//nbins)
        bin_edges[0,range(nbins,nbins*2)]=np.arange(POT_1,POT_2,POT_1//nbins)
        bin_edges[0,range(2*nbins,nbins*3+1)]=np.arange(POT_2,Tr+1,POT_1//nbins) 
  
        for bins in range(nbins*3):
            binned_Errorscore[0,bins]=np.mean(np.mean(Errorscore[:,range(int(bin_edges[0,bins]),int(bin_edges[0,bins+1]))],axis=1))
            binned_accuracy[0,bins]=np.mean(rew[0,range(int(bin_edges[0,bins]),int(bin_edges[0,bins+1]))])

        data_pickle = "Beta"+str(b)+"_Rep"+str(r)+"_RBM_nosync.p"

        variables_str = ["binned_Errorscore","binned_accuracy"]
        variables =     [binned_Errorscore, binned_accuracy]
        data = []
        for loop in range(len(variables)):
            data.append({variables_str[loop]: variables[loop]})
            #print(data)
            pickle.dump(data,open(data_pickle,"wb"))
            #myworkspacetoo = pickle.load(open(data_pickle, "rb"))
            #print(myworkspacetoo)